### PR TITLE
[Linux] Collect Linux OVA configs at OVA deployment

### LINKED ISSUE #418: [New Feature] Collect OVF configs for Linux OVA deploy
##### SUMMARY

This feature is to add OVF configs collection in deploy_vm_from_ova.yml

##### ISSUE TYPE
- Feature Idea

### PR DESCRIPTION
**Pull request by keirazhang**
Fix #418 
Changes include:
1. Collected CDROMs, USBs, Video card information in `common/collect_ovf_vm_config.yml`
2. Tasks in `linux/deploy_vm/amazonlinux/reconfigure_amazonlinux_vm.yml` and `linux/deploy_vm/photon/reconfigure_photon_vm.yml` are almost same except for the variables for create seed ISO. And Ubuntu OVA deployment also includes these tasks in `linux/deploy_vm/ubuntu/reconfigure_ubuntu_ova_vm.yml`, except that Ubuntu OVA needs an extra serial port and CDROM. So here removed redundant tasks files and use a united tasks file `linux/deploy_vm/reconfigure_vm_with_cloudinit.yml`.
3. Amazon Linux post configs were moved into its cloud-init user data, which can be completed at cloud-init config.

### CODE REVIEW

**Comment by keirazhang** [comment]
```

  - 'Guest os distribution: VMware Photon OS 4.0 x86_64'
  - 'Hardware version: vmx-20'
  - 'Config guest id: vmwarePhoton64Guest'
  - 'Guestinfo guest id: vmwarePhoton64Guest'
  - 'Guestinfo guest full name: VMware Photon OS (64-bit)'
  - 'Guestinfo guest family: linuxGuest'
  - 'Guestinfo detailed data: architecture=''X86'' bitness=''64'' distroName=''VMware
    Photon OS'' distroVersion=''4.0'' familyName=''Linux'' kernelVersion=''5.10.83-6.ph4-esx''
    prettyName=''VMware Photon OS/Linux'''
  - 'Results: Total (2), Passed (2)'
  - ' * deploy_photon_ova...............................................Passed'
  - ' * check_inbox_driver..............................................Passed'

  - 'Guest os distribution: Amazon 2 x86_64'
  - 'Hardware version: vmx-20'
  - 'Config guest id: other3xLinux64Guest'
  - 'Guestinfo guest id: amazonlinux2_64Guest'
  - 'Guestinfo guest full name: Amazon Linux 2 (64-bit)'
  - 'Guestinfo guest family: linuxGuest'
  - 'Guestinfo detailed data: architecture=''X86'' bitness=''64'' distroName=''Amazon
    Linux'' distroVersion=''2'' familyName=''Linux'' kernelVersion=''4.14.301-224.520.amzn2.x86_64''
    prettyName=''Amazon Linux 2'''
  - 'Results: Total (2), Passed (2)'
  - ' * deploy_amazon_ova...............................................Passed'
  - ' * check_inbox_driver..............................................Passed'

  - 'Guest os distribution: Flatcar 3374.2.3 x86_64'
  - 'Hardware version: vmx-20'
  - 'Config guest id: other3xLinux64Guest'
  - 'Guestinfo guest id: other5xLinux64Guest'
  - 'Guestinfo guest full name: Other 5.x Linux (64-bit)'
  - 'Guestinfo guest family: linuxGuest'
  - 'Guestinfo detailed data: architecture=''X86'' bitness=''64'' buildNumber=''2023-01-20-1751''
    distroName=''Flatcar Container Linux by Kinvolk'' distroVersion=''3374.2.3'' familyName=''Linux''
    kernelVersion=''5.15.86-flatcar'' prettyName=''Flatcar Container Linux by Kinvolk
    3374.2.3 (Oklo)'''
  - 'Results: Total (2), Passed (2)'
  - ' * deploy_flatcar_ova..............................................Passed'
  - ' * check_inbox_driver..............................................Passed'

  - 'Guest os distribution: Ubuntu 22.04 x86_64'
  - 'Hardware version: vmx-20'
  - 'Config guest id: ubuntu64Guest'
  - 'Guestinfo guest id: ubuntu64Guest'
  - 'Guestinfo guest full name: Ubuntu Linux (64-bit)'
  - 'Guestinfo guest family: linuxGuest'
  - 'Guestinfo detailed data: architecture=''X86'' bitness=''64'' distroName=''Ubuntu
    22.04 LTS'' distroVersion=''22.04'' familyName=''Linux'' kernelVersion=''5.15.0-25-generic''
    prettyName=''Ubuntu 22.04 LTS'''
  - 'Results: Total (2), Passed (2)'
  - ' * deploy_ubuntu_ova...............................................Passed'
  - ' * check_inbox_driver..............................................Passed'

  - 'Guest os distribution: Ubuntu 22.10 x86_64'
  - 'Hardware version: vmx-20'
  - 'Config guest id: ubuntu64Guest'
  - 'Guestinfo guest id: ubuntu64Guest'
  - 'Guestinfo guest full name: Ubuntu Linux (64-bit)'
  - 'Guestinfo guest family: linuxGuest'
  - 'Guestinfo detailed data: architecture=''X86'' bitness=''64'' distroName=''Ubuntu
    22.10'' distroVersion=''22.10'' familyName=''Linux'' kernelVersion=''5.19.0-21-generic''
    prettyName=''Ubuntu 22.10'''
  - 'Results: Total (2), Passed (2)'
  - ' * deploy_ubuntu_ova...............................................Passed'
  - ' * check_inbox_driver..............................................Passed'

  - 'Guest os distribution: Ubuntu 20.04 x86_64'
  - 'Hardware version: vmx-20'
  - 'Config guest id: ubuntu64Guest'
  - 'Guestinfo guest id: ubuntu64Guest'
  - 'Guestinfo guest full name: Ubuntu Linux (64-bit)'
  - 'Guestinfo guest family: linuxGuest'
  - 'Guestinfo detailed data: architecture=''X86'' bitness=''64'' distroName=''Ubuntu''
    distroVersion=''20.04'' familyName=''Linux'' kernelVersion=''5.4.0-125-generic''
    prettyName=''Ubuntu 20.04.4 LTS'''
  - 'Results: Total (2), Passed (2)'
  - ' * deploy_ubuntu_ova...............................................Passed'
  - ' * check_inbox_driver..............................................Passed'
```
Amazon Linux 2
```
[
    {
        "CDROMs": [],
        "CPU Cores per Socket": 1,
        "CPU Hot Add Enabled": true,
        "CPU Number": 2,
        "Disks": [
            {
                "Controller": "VMware paravirtual SCSI",
                "Disk Size in GB": 25.0,
                "Label": "Hard disk 1",
                "Thin Provisioned": true
            }
        ],
        "Firmware": "bios",
        "Guest OS Distribution": "Amazon Linux 2",
        "Guest OS ID": "other3xLinux64Guest",
        "Guest OS Version": "Other 3.x or later Linux (64-bit)",
        "Hardware Version": "vmx-11",
        "Kernel Version": "4.14.301-224.520.amzn2.x86_64",
        "Memory Hot Add Enabled": true,
        "Memory Size in MB": 4096,
        "Nested HV Enabled": false,
        "Network Adapters": [
            {
                "Label": "Network adapter 1",
                "Type": "VMXNET3"
            }
        ],
        "Secure Boot": false,
        "USB Controllers": [
            "USB controller"
        ],
        "VBS Enabled": false,
        "VMware Tools": {
            "Install Type": "guestToolsTypeOpenVMTools",
            "Sync Time with Host": true,
            "Upgrade Policy": "manual",
            "Version": "12.1.0 (build-20219665)"
        },
        "Video Card": {
            "3D Memory Size in MB": 256.0,
            "3D Renderer": "automatic",
            "Auto-detect Settings": false,
            "Enable 3D Support": false,
            "Number of Displays": 1,
            "Video RAM Size in MB": 4.0
        },
        "vTPM": false
    }
]
```
Flatcar
```
[
    {
        "CDROMs": [
            {
                "Controller": "IDE 1",
                "Label": "CD/DVD drive 1"
            }
        ],
        "CPU Cores per Socket": 1,
        "CPU Hot Add Enabled": true,
        "CPU Number": 2,
        "Disks": [
            {
                "Controller": "VMware paravirtual SCSI",
                "Disk Size in GB": 8.490234375,
                "Label": "Hard disk 1",
                "Thin Provisioned": true
            }
        ],
        "Firmware": "bios",
        "Guest OS Distribution": "Flatcar Container Linux by Kinvolk 3374.2.3 (Oklo)",
        "Guest OS ID": "other3xLinux64Guest",
        "Guest OS Version": "Other 3.x or later Linux (64-bit)",
        "Hardware Version": "vmx-13",
        "Kernel Version": "5.15.86-flatcar",
        "Memory Hot Add Enabled": true,
        "Memory Size in MB": 1024,
        "Nested HV Enabled": false,
        "Network Adapters": [
            {
                "Label": "Network adapter 1",
                "Type": "VMXNET3"
            }
        ],
        "Secure Boot": false,
        "USB Controllers": [],
        "VBS Enabled": false,
        "VMware Tools": {
            "Install Type": "guestToolsTypeOpenVMTools",
            "Sync Time with Host": false,
            "Upgrade Policy": "manual",
            "Version": "12.1.0 (build-20219665)"
        },
        "Video Card": {
            "3D Memory Size in MB": 256.0,
            "3D Renderer": "automatic",
            "Auto-detect Settings": false,
            "Enable 3D Support": false,
            "Number of Displays": 1,
            "Video RAM Size in MB": 4.0
        },
        "vTPM": false
    }
]
```
Windows
```
[
    {
        "CDROMs": [
            {
                "Controller": "IDE 1",
                "Label": "CD/DVD drive 1"
            },
            {
                "Controller": "IDE 0",
                "Label": "CD/DVD drive 2"
            }
        ],
        "CPU Cores per Socket": 1,
        "CPU Hot Add Enabled": false,
        "CPU Number": 4,
        "Disks": [
            {
                "Controller": "LSI Logic SAS",
                "Disk Size in GB": 125.0,
                "Label": "Hard disk 1",
                "Thin Provisioned": true
            }
        ],
        "Firmware": "efi",
        "Guest OS Distribution": "Windows 10 Enterprise, 64-bit (Build 22621.1105)",
        "Guest OS ID": "windows9_64Guest",
        "Guest OS Version": "Microsoft Windows 10 (64-bit)",
        "Hardware Version": "vmx-19",
        "Memory Hot Add Enabled": false,
        "Memory Size in MB": 4096,
        "Nested HV Enabled": true,
        "Network Adapters": [
            {
                "Label": "Network adapter 1",
                "Type": "E1000"
            },
            {
                "Label": "Network adapter 2",
                "Type": "E1000"
            }
        ],
        "Secure Boot": true,
        "USB Controllers": [],
        "VBS Enabled": false,
        "VMware Tools": {
            "Install Type": "guestToolsTypeUnknown",
            "Sync Time with Host": true,
            "Upgrade Policy": "upgradeAtPowerCycle",
            "Version": "11.3.5 (build-18557794)"
        },
        "Video Card": {
            "3D Memory Size in MB": 256.0,
            "3D Renderer": "automatic",
            "Auto-detect Settings": false,
            "Enable 3D Support": false,
            "Number of Displays": 1,
            "Video RAM Size in MB": 256.0
        },
        "vTPM": false
    }
]
```

File: linux/deploy_vm/reconfigure_vm_with_cloudinit.yml [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+  include_tasks: ../../common/vm_get_cdrom_devices.yml
+
+- name: "Set fact of VM's existing CDROM device"
+  ansible.builtin.set_fact:
+    vm_existing_cdrom_list: "{{ cdrom_device_list }}"
+
+# Amazon Linux and VMware Photon OS VM can use existing CDROM device
+- name: "Use existing CDROM device for attaching cloud-init seed ISO"
+  ansible.builtin.set_fact:
+    vm_cdrom_controller_type: "{{ vm_existing_cdrom_list[0].controller_label.split()[0] |
```
Comment (Tomorrow9) [inline linux/deploy_vm/reconfigure_vm_with_cloudinit.yml:50]: It seems this 'default' value is not needed.

File: linux/deploy_vm/reconfigure_vm_with_cloudinit.yml [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+                                 lower | default('sata') }}"
+    vm_cdrom_controller_num: "{{ vm_existing_cdrom_list[0].bus_num }}"
+    vm_cdrom_unit_num: "{{ vm_existing_cdrom_list[0].unit_num }}"
+  when:
+    - ova_guest_os_type in ['photon', 'amazon']
```
Comment (Tomorrow9) [inline linux/deploy_vm/reconfigure_vm_with_cloudinit.yml:55]: Can this condition `ova_guest_os_type in ['photon', 'amazon']` be removed to fit common situation.

File: linux/deploy_vm/deploy_vm_from_ova.yml [hunk 3]
```diff
@@ -51,88 +49,76 @@
...
-    - include_tasks: ubuntu/reconfigure_ubuntu_ova_vm.yml
-      when: ova_guest_os_type == 'ubuntu'
-
-    # Warning about other OS OVAs which are not supported
-    - ansible.builtin.debug:
-        msg: "WARN: OVA guest OS type is not supported in this project. The following tests might fail."
+    - name: "Warning about unknown guest OS type"
+      ansible.builtin.debug:
+        msg: >-
+          WARNING: The guest OS type of the OVA is not supported in this project.
```
Comment (Tomorrow9) [inline linux/deploy_vm/deploy_vm_from_ova.yml:102]: I think here can be changed to "not matched known guest OS type reconfiguration file/method, please add it if needed or following tests might fail."

File: linux/deploy_vm/reconfigure_vm_with_cloudinit.yml (hunk 2 above)
Comment (keirazhang) [reply linux/deploy_vm/reconfigure_vm_with_cloudinit.yml:55]: At present, only Ubuntu, Photon, Amazon Linux will enter `reconfigure_vm_with_cloudinit.yml`. The other OVA with unknown guest OS will go straight tasks after `reconfigure_vm_with_cloudinit.yml`.

File: linux/deploy_vm/reconfigure_vm_with_cloudinit.yml (hunk 1 above)
Comment (keirazhang) [reply linux/deploy_vm/reconfigure_vm_with_cloudinit.yml]: Removed 'default' value.

File: linux/deploy_vm/reconfigure_vm_with_cloudinit.yml (hunk 2 above)
Comment (keirazhang) [reply linux/deploy_vm/reconfigure_vm_with_cloudinit.yml:54]: For other guest OS type we can suppose it doesn't need any reconfiguration for tests unless there is bug or new feature request for it.

File: linux/deploy_vm/deploy_vm_from_ova.yml (hunk 3 above)
Comment (keirazhang) [reply linux/deploy_vm/deploy_vm_from_ova.yml]: Message is updated.

**Review by Tomorrow9 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**